2-WebApp-graph-user/2-5-HybridFlow/Views/Shared/_Layout.cshtml

```diff
@@ -164,6 +155,28 @@
                 {
                     @Html.Raw($"const code = '';");
                 }
+
+                if (User.Identity is not null)
+                {
+                    @Html.Raw($"const username = '{User.Identity.Name}';");
+                }
+                else
+                {
+                    @Html.Raw($"const username = '';");
+                }
+            }
+
+            @{ /* Remove any client sessions if user is not logged in server-side or a different user is logged in. */ }
+            if (!!msalInstance.getAllAccounts().length) {
+                if(!username ||
+                   !msalInstance.getAllAccounts()
+                        .find(a => a.username === username)) {
+
+                    msalInstance.logoutRedirect({
+                        onRedirectNavigate: _ => false
+                    });
+
+                }
```

Comment:
@jasonnutter 

Would this work?

On the front-end, this runs each time any page in the application loads. The `username` value is generated server-side. If no session is active the value is empty and triggers the logout without redirect to empty the client cache.

Reply:
I would recommend to show an error message (and/or prompt to login), instead of auto logging out.

Reply:
Also, make sure to not include this logic on pages used as the MSAL redirect URI, as you can introduce a race condition that could prevent users from logging in.

Reply:
The idea here is that this would be treated as part of the main logout flow. The error handling is incidental.

**ASP.NET Core** has an out of the box `SignedOut` page that it uses when you use the pre-configured logout flow. You _can_ overwrite it to add a custom page but my point of resistance is it's a touch messy code-wise.

Here, instead of writing a custom page the central page would be configured to remove any sessions when users are not logged in server-side.

Barring errors, msal.js authenticates the user exclusively with an auth code. The only time a user would be prompted is if there was an error client-side, e.g. trying to redeem a bad auth code.

Reply:
I guess I'm confused, because it looks like this code runs on every page, and it looks like you are checking if teh user is logged in before MSAL has redeemed the auth code (i.e. you are seeing if they have a cached account in MSAL before you complete the process of logging in which puts accounts in the cache).

Reply:
My point is that you should only call a logout API when teh user indicated they want to logout (or your app has received a logout signal from another app via front-channel logout).

Reply:
@jasonnutter that is correct and this page _does_ run everywhere but the the logic is rendered server-side.

What happens is the server checks that there is a valid session for the user. If there is a valid session for the user and an auth code exists in the session then the auth code is redeemed. If the user has an token available in the cache then there will be no code to redeem and the client will use the cached token.

The idea here is that it's supposed to work as a sort of failsafe. If, for whatever reason, there is no session for the user on the server and a client-side token exists that token will be discarded. The user will then need to login again to get a new access token in the client.

I get what you're saying about calling `logout`. It is a bit confusing. Is there another way to dump tokens in the client without calling the logout API?

Reply:
To try and add a touch more clarity here.

In these samples there are several out of the box features that have pre-configured and built endpoints including one endpoint for managing log out.

In the `_LoginPartial.cshtml` you'll find the tag that renders the `Sign Out` button written as

```
<a class="nav-link text-dark" asp-area="MicrosoftIdentity" asp-controller="Account" asp-action="SignOut">Sign out</a>
```

After it's rendered server-side this is the link that's generated.

```
<a class="nav-link text-dark" href="/MicrosoftIdentity/Account/SignOut">Sign out</a>
```

It's certainly possible to add an `onClick` attribute to run the proper JavaScript and continue to the sign-out link that way. Ideally it shouldn't matter but I want to ensure I'm leveraging as many of the 'out of the box' features here as I can.

Reply:
After playing around with this I decided to meet in the middle here.

I remove the `logout` calls in the main page and now do it exclusively in the `_LoginPartial.cshtml` page.

Added `click` handler to the link and use the URL generated by ASP.NET.